src/main/java/org/apache/commons/dbcp2/PoolingDataSource.java

```diff
@@ -77,6 +77,9 @@ public PoolingDataSource(final ObjectPool<C> pool) {
     @Override
     public void close() throws RuntimeException, SQLException {
         try {
+            if (pool instanceof GenericObjectPool) {
+                ((PoolableConnectionFactory) ((GenericObjectPool<?>) pool).getFactory()).close();
```

Comment:
Should the cast to `PoolableConnectionFactory` be changed `AutoCloseable` for a touch more flexibility?

Reply:
+1